kv/txn_coord_sender.go

```diff
@@ -131,33 +131,41 @@ func (tm *txnMetadata) close(txn *proto.Transaction, resolved []proto.Key, sende
 		}
 	}
 	for _, o := range tm.keys.GetOverlaps(engine.KeyMin, engine.KeyMax) {
-		call := client.Call{
-			Args: &proto.InternalResolveIntentRequest{
+		// If the op was range based, end key != start key: resolve a range.
+		var call client.Call
+		key := o.Key.Start().(proto.Key)
+		endKey := o.Key.End().(proto.Key)
+		if !key.Next().Equal(endKey) {
+			call.Args = &proto.InternalResolveIntentRangeRequest{
 				RequestHeader: proto.RequestHeader{
 					Timestamp: txn.Timestamp,
-					Key:       o.Key.Start().(proto.Key),
+					Key:       key,
+					EndKey:    endKey,
 					User:      storage.UserRoot,
 					Txn:       txn,
 				},
-			},
-			Reply: &proto.InternalResolveIntentResponse{},
-		}
-		// Set the end key only if it's not equal to Key.Next(). This
-		// saves us from unnecessarily clearing intents as a range.
-		endKey := o.Key.End().(proto.Key)
-		if !call.Args.Header().Key.Next().Equal(endKey) {
-			call.Args.Header().EndKey = endKey
+			}
+			call.Reply = &proto.InternalResolveIntentRangeResponse{}
 		} else {
 			// Check if the key has already been resolved; skip if yes.
 			found := false
 			for _, k := range resolved {
-				if call.Args.Header().Key.Equal(k) {
+				if key.Equal(k) {
 					found = true
 				}
 			}
 			if found {
 				continue
 			}
+			call.Args = &proto.InternalResolveIntentRequest{
```

Comment:
That's what I did first. But other complications ensued. In the end I reverted that experiment and did this because even though it's one more rpc type (who cares really?), it properly tracks with the idea that ops are either range-based or not.